### PR TITLE
Implement TypeVar(..., bound=<type>)

### LINKED ISSUE #689: Implement TypeVar(..., bound=<boundary_type>)
Now specified in PEP 484; see https://github.com/ambv/typehinting/issues/59.

### PR DESCRIPTION
**Pull request by rwbarton**
Fixes #689.

This will also need documentation, would you like me to write it and add it to this pull request?

### CODE REVIEW

**Comment by rwbarton** [comment]
Added a missing `return None` in `process_typevar_parameters`.

File: mypy/test/data/check-bound.test [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+[case testBoundMethodUsage]
+from typing import TypeVar
+class A:
+    def foo(self) -> None: pass
+class B(A): pass
+
+T = TypeVar('T', A)
+
+def f(x: T) -> T:
+    x.foo()
```
Comment (JukkaL) [inline mypy/test/data/check-bound.test]: Test also attribute, property, static method and class method. Also test inherited method.

File: mypy/test/data/check-bound.test [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+
+[case testBoundHigherOrderWithVoid]
+from typing import TypeVar, Callable
+class A: pass
+T = TypeVar('T', bound=A)
+def f(g: Callable[[], T]) -> T:
+    return g()
+def h() -> None: pass
+f(h)
```
Comment (JukkaL) [inline mypy/test/data/check-bound.test]: Verify that mypy complains about using the result (a `None` return value can't be used as value).

File: mypy/test/data/check-bound.test [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+[out]
+main:10: error: Type argument "builtins.str" of "G" must be a subtype of "__main__.A"
+main:11: error: Type argument 1 of "G" has incompatible value "str"
+
+
+[case testBoundVoid]
+from typing import TypeVar, Generic
+T = TypeVar('T', bound=int)
+class C(Generic[T]): pass
+c1 = None # type: C[None]
```
Comment (JukkaL) [inline mypy/test/data/check-bound.test]: Have a method that returns `T` in `C` and try calling that and using the return value.

File: mypy/test/data/check-bound.test [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+from typing import TypeVar, Generic
+
+class A: pass
+class B(A): pass
+T = TypeVar('T', bound=A)
+
+class G(Generic[T]):
+    def __init__(self, x: T) -> None: pass
+
+x = None # type: G[str] # Fail
```
Comment (JukkaL) [inline mypy/test/data/check-bound.test]: Test that `G[A]` and `G[B]` are valid.

File: mypy/test/data/check-bound.test [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+class B(A): pass
+
+T = TypeVar('T')
+TA = TypeVar('TA', bound=A)
+TB = TypeVar('TB', bound=B)
+
+def f(x: T) -> T:
+    return x
+def g(x: TA) -> TA:
+    return f(x)
```
Comment (JukkaL) [inline mypy/test/data/check-bound.test]: Also check that `TA` is a subtype of `A` but not a subtype of `B`.

File: mypy/test/data/semanal-errors.test [hunk 6]
```diff
@@ -1005,11 +1005,12 @@ d = TypeVar('D')    # E: Unexpected TypeVar() argument value
 e = TypeVar('e', int, str, x=1) # E: Unexpected argument to TypeVar(): x
 f = TypeVar('f', (int, str)) # E: Type expected
 g = TypeVar('g', x=(int, str)) # E: Unexpected argument to TypeVar(): x
+h = TypeVar('h', bound=1) # E: TypeVar 'bound' must be a type
 [out]
 
 [case testMoreInvalidTypevarArguments]
 from typing import TypeVar
-T = TypeVar('T', bound=int) # E: TypeVar 'bound' argument not supported yet
+T = TypeVar('T', int, str, bound=bool) # E: TypeVar cannot have both values and an upper bound
```
Comment (JukkaL) [inline mypy/test/data/semanal-errors.test]: Using `bound=` and `contravariant=True` should likely be an error?

**Comment by JukkaL** [comment]
Looks good! Mostly some suggestions for additional test cases above.

Some other things:

* `visit_type_var_expr` in `strconv.py` doesn't consider `upper_bound`. This can be tested via `semanal-*.test`.
* I wonder if it would be useful to have a test case for `bound=` and `covariant=True`?
* I'd like some tests for inheritance with a generic class and bound. For example, if base class has a bound, mypy needs to enforce that a subclass also has a compatible bound:

    ```
    TA = TypeVar('TA', bound=A)
    T = TypeVar('T')

    class C(Generic[TA]): ...
    class D(C[T], Generic[T]): ...  # should be an error
    ```

* Another inheritance example (worth adding a test case, likely works already):

    ```
    class A: pass
    T = TypeVar('T', bound=A)

    class C(Generic[T]): ...
    class D(C[A]): ...  # okay
    class D(C[str]): ...  # error
    ```
    
If any of my suggestions are tricky to implement, feel free to create issues for them and we can postpone them until after this has landed -- this already unblocks a lot of things.

File: mypy/test/data/check-bound.test [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+f(g())
+C(g())
+z = None # type: C
+
+
+[case testBoundHigherOrderWithVoid]
+from typing import TypeVar, Callable
+class A: pass
+T = TypeVar('T', bound=A)
+def f(g: Callable[[], T]) -> T:
```
Comment (JukkaL) [inline mypy/test/data/check-bound.test]: Test a decorator with a bound of `Callable[..., Any]`, as users have asked about this before, I think.

**Comment by rwbarton** [comment]
Thanks for all the test suggestions! I've added most of them.

I'm inclined to leave the interaction of variance and bounds to another ticket, since the first thing to do there is figure out exactly what we should do, and I haven't been thinking about variance at all yet.

> visit_type_var_expr in strconv.py doesn't consider upper_bound. This can be tested via semanal-*.test

Probably `TypeVarDef.__repr__` and `TypeStrVisitor` should display the upper bound too, for debugging purposes.

**Comment by JukkaL** [comment]
I did a quick pass and this looks good to merge. We can leave other things for later, as they likely won't have much user impact yet.

**Comment by gvanrossum** [comment]
I am testing this with a callable bound, and I get an unexpected error. Example:
```
from typing import *

F = TypeVar('F', bound=Callable[..., Any])

def my_decorator(f):
    # type: (F) -> F

    def inner(*args):
        # type: (*Any) -> Any
        return f(*args)   # <---------- error here

    return cast(F, inner)
```
I get the error:
```
t.py: note: In function "inner":
t.py:10: error: "F" not callable
```
I also wish I wouldn't need the `cast(F, inner)` call but I can sort of understand that the types aren't quite compatible, so I can live with it.

**Comment by rwbarton** [comment]
> I get the error:
> 
> t.py: note: In function "inner":
> t.py:10: error: "F" not callable

Ah! Good catch. I hadn't realized that an unannotated inner function inside an annotated outer function is not type checked, so my test that is very similar to your example wasn't testing what I thought it was, and in fact this doesn't work for some reason. I'll investigate.

> I also wish I wouldn't need the `cast(F, inner)` call but I can sort of understand that the types aren't quite compatible, so I can live with it.

Yeah. You can't declare `inner` as having the type `F`, even using the `# type:` syntax; the mypy parser expects a type that looks syntactically like a function. Even if you could, mypy probably wouldn't be able to check the body of `inner` against the type variable `F` anyways.

A variadic type variable would be better here, and then you wouldn't actually need `bound=...`: you could just write `def my_decorator(f: Callable[A, B]) -> Callable[A, B]: ...`, where `A` can range over appropriate argument lists.

**Comment by gvanrossum** [comment]
I definitely hope you can fix the first issue, since all three places in
our code where we want `bound=` use this exact idiom. :-)

We'll get to variadic type variables later.

**Comment by JukkaL** [comment]
One more thing that would be good to test is overloading + bounds. Example:

```
T = TypeVar('T', bound=str)

@overload
def f(x: T) -> T: ...
@overload
def f(x: int) -> int: ...
```

Now `f('x')` should have type `str`, while `f(1)` should have type `int`. Also consider subclasses of `str`. `f(1.1)` should be an error.

**Comment by rwbarton** [comment]
Indeed. I noticed this too while looking at #1241. Please make sure you agree with the behavior of the new tests as I'm not sure I yet have the correct understanding of how overloading is supposed to be resolved.

**Comment by gvanrossum** [comment]
I re-checked this with my primary test case and it works! You need to rebase though, and I'll let Jukka do the final review.

**Comment by JukkaL** [comment]
This is now good to merge!

**Comment by rwbarton** [comment]
Rebased on master

**Pull request merged**